cai_causal_graph/identify_utils.py

```diff
@@ -92,30 +122,138 @@ def _identify_confounders_no_checks_no_descendant_pruning(clean_graph: CausalGra
+
+    # reject all candidate nodes that have a directed path to the destination that does not go through the source
+    for candidate in candidate_nodes.copy():
+        for causal_path in graph.get_all_causal_paths(candidate, destination_id):
+            if source_id not in causal_path:
+                candidate_nodes.remove(candidate)
+                break
+
+    # reject all candidate nodes that have a non-empty confounding set between itself and the destination
+    for candidate in candidate_nodes.copy():
+        if len(identify_confounders(graph, candidate, destination_id)) > 0:
+            candidate_nodes.remove(candidate)
+
+    return list(candidate_nodes)
+
+
+def identify_mediators(graph: CausalGraph, source: NodeLike, destination: NodeLike) -> List[str]:
+    """
+    Identify all mediators for the causal effect of `source` and `destination` in the provided `graph`.
+
+    A mediator variable for the causal effect of `source` on `destination` satisfies the following criteria:
+        1. There is a causal effect between the `source` and the `mediator`.
+        2. There is a causal effect between the `mediator` and the `destination`.
+        3. The `mediator` blocks all directed causal paths between the `source` and the `destination`.
+        4. There is no directed causal path from any confounder between `source` and `destination` to the `mediator`.
+
+    Example:
+        >>> from typing import List
+        >>> from cai_causal_graph import CausalGraph
+        >>> from cai_causal_graph.identify_utils import identify_mediators
+        >>>
+        >>> # define a causal graph
+        >>> cg = CausalGraph()
+        >>> cg.add_edge('x', 'm')
+        >>> cg.add_edge('m', 'y')
+        >>> cg.add_edge('u', 'x')
+        >>> cg.add_edge('u', 'y')
+        >>> cg.add_edge('x', 'y')
+        >>>
+        >>> # find the mediators between 'x' and 'y'; output: ['m']
+        >>> mediator_variables: List[str] = identify_mediators(cg, source='x', destination='y')
+
+    :param graph: The causal graph given by a `cai_causal_graph.causal_graph.CausalGraph` instance. This must be a DAG,
+        i.e. it must only contain directed edges and be acyclic, otherwise a `TypeError` is raised.
+    :param source: The source variable or its identifier.
+    :param destination: The destination variable or its identifier.
+    :return: A list of mediator variables for the causal effect of `source` on `destination`.
+    """
+    # verify inputs and obtain node identifiers
+    source_id, destination_id = _verify_identify_inputs(graph, source, destination)
+
+    # get all the confounders between the source and destination
+    confounders = identify_confounders(graph, source_id, destination_id)
+
+    # query all causal paths between the source and destination (also remove trivial paths)
+    causal_paths = [set(path) for path in graph.get_all_causal_paths(source_id, destination_id) if len(path) > 2]
```

Comment:
Would maybe add a max_num_paths and raise if more than that. Otherwise this may run forever. Don't think we can stop early as the list of mediators may be bad. We can default `max_num_paths` to 25 or something. 
```
causal_paths = []
for i, path in enumerate(graph.get_all_causal_paths(source_id, destination_id)):
   if i > max_num_paths:
      # raise nice error
   elif len(path) > 2:
      causal_paths.append(set(path))
   # No else needed as path is just source -> destination
```

Reply:
Guess this may apply for instruments too. Just a little nervous whenever all paths are requested as this scales poorly.

Reply:
That's a good suggestion, thank you. I agree with not being able to stop early and raising an error instead. Will add to both instruments and mediators.